src/main/java/org/elasticsearch/search/aggregations/bucket/nested/NestedAggregator.java

```diff
@@ -175,4 +177,78 @@ public InternalAggregation buildEmptyAggregation() {
             }
         }
     }
+
+    // The aggs framework can collect buckets for the same parent doc id more than once and because the children docs
+    // can only be consumed once we need to buffer the child docs. We only need to buffer child docs in the scope
+    // of the current root doc.
+
+    // Examples:
+    // 1) nested agg wrapped is by terms agg and multiple buckets per document are emitted
+    // 2) Multiple nested fields are defined. A nested agg joins back to another nested agg via the reverse_nested agg.
+    //      For each child in the first nested agg the second nested agg gets invoked with the same buckets / docids
+    private ChildDocIterator getChildIterator(final int parentDocId) throws IOException {
+        int rootDocId = rootDocs.nextSetBit(parentDocId);
+        if (currentRootDoc == rootDocId) {
+            final IntArrayList childDocIdBuffer = childDocIdBuffers.get(parentDocId);
+            if (childDocIdBuffer != null) {
+                return new ChildDocIterator() {
```

Comment:
Could childDocIdBuffers store a reusable iterator? Would avoid the object creation overhead every time a parent is re-collected.